libtenzir/builtins/operators/cache.cpp

```diff
@@ -84,131 +95,184 @@ struct cache_state {
     });
   }
 
   auto set_write_timeout() -> void {
-    if (write_timeout == duration::zero()) {
+    if (write_timeout_ == duration::zero()) {
       return;
     }
-    TENZIR_ASSERT(write_timeout > duration::zero());
-    on_write_timeout.dispose();
-    on_write_timeout = detail::weak_run_delayed(self, write_timeout, [this] {
-      self->quit(diagnostic::error("cache expired").to_error());
+    TENZIR_ASSERT(write_timeout_ > duration::zero());
+    on_write_timeout_.dispose();
+    on_write_timeout_ = self_->run_delayed_weak(write_timeout_, [this] {
+      self_->quit(diagnostic::error("cache expired").to_error());
     });
   }
 
   auto write_ok() -> caf::result<bool> {
-    return writer;
+    return writer_;
   }
 
   auto write(table_slice events) -> caf::result<bool> {
     TENZIR_ASSERT(events.rows() > 0);
-    const auto sender = self->current_sender();
+    const auto sender = self_->current_sender();
     TENZIR_ASSERT(sender);
-    if (not writer) {
-      writer = sender->address();
-      self->monitor(sender, [this](const caf::error&) {
-        TENZIR_ASSERT(not done);
-        done = true;
+    if (not writer_) {
+      writer_ = sender->address();
+      self_->monitor(sender, [this](const caf::error& err) {
+        TENZIR_ASSERT(not done_);
+        done_ = true;
+        update_multicaster_.push(cache_update{
+          .approx_bytes = byte_size_,
+          .state = err ? cache_state::failed : cache_state::closed,
+        });
+        update_multicaster_.close();
         reset_read_timeout();
         // We ignore error messages because they do not matter to the readers.
-        for (auto& [_, reader] : readers) {
-          if (reader.offset == cache.size() and reader.rp.pending()) {
+        for (auto& [_, reader] : readers_) {
+          if (reader.offset == cached_events_.size() and reader.rp.pending()) {
             reader.rp.deliver(table_slice{});
           }
         }
       });
       set_write_timeout();
-    } else if (writer != sender->address()) {
+    } else if (writer_ != sender->address()) {
       return false;
     }
     auto exceeded_capacity = false;
-    if (cache_size + events.rows() > capacity.inner) {
-      events = head(std::move(events), capacity.inner - cache_size);
+    if (cache_size_ + events.rows() > capacity_.inner) {
+      events = head(std::move(events), capacity_.inner - cache_size_);
       diagnostic::warning("cache exceeded capacity")
-        .primary(capacity.source)
-        .emit(diagnostics);
+        .primary(capacity_.source)
+        .emit(diagnostics_);
       exceeded_capacity = true;
       if (events.rows() == 0) {
         return false;
       }
     }
-    cache_size += events.rows();
-    cache.push_back(std::move(events));
-    for (auto& [_, reader] : readers) {
+    cache_size_ += events.rows();
+    // Calculate the byte size of any referenced Arrow buffers
+    // TODO: Use `table_slice::approx_bytes` instead.
```

Comment:
I'm not sure this TODO makes sense here, as the cache will still keep to full record batch in memory instead of only the data, not the sliced one.

Reply:
This comment is just there due to a conflict with #4981, which introduces a custom mechanism that is faster than Arrow's mechanism for measuring approx. bytes for a record batch. Ideally I would want to use that here already. Depending on which PR goes in first I'll make use of it, or I'll open another PR later for it.

Reply:
IIRC `approx_bytes` was used to measure the byte size of a slice that was backed by a larger record batch + a mask, did that change?.

Reply:
Not at all, it just got moved into an optimized implementation we do ourselves. The calculation is the same that the Arrow function I'm calling here is doing, just without allocations.

Reply:
I didn't know `ReferencedBufferSize` does the same thing. Still, I'm not convinced it the approximate value should be used here, as it makes the meaning of `tenzir.cache.capacity` very fuzzy.

Reply:
The value is an approximation no matter what, though. If we go through the `where` operator and that yields every second row, we would massively overcount if we used `TotalBufferSize` instead of `ReferencedBufferSize`.

Reply:
> ... we would massively overcount if we used `TotalBufferSize` instead of `ReferencedBufferSize`.

In terms of data that is being kept in memory `TotalBufferSize` gives the correct value. Since we have no option to limit bytes at the operator instance there is no conflict with the "virtual" view.

Reply:
That's not true once you have more than one table slice. If they share memory, then `TotalBufferSize` is overestimating by a very large margin. Since this happens a lot more frequently in our operators than the case you're describing I'd be hesitant to change this.

Reply:
I clarified the behavior in /70fc1b965cbbdce3d5caed9fd43bc619372c92e2

Reply:
We documented that the existing behavior now.